create order orchestrator with camunda

MSurmach opened the pull request.

MSurmach committed:
create order orchestrator with camunda

MSurmach committed:
bug fixes and improvements

Tyomich4736 commented on order-service/src/main/java/com/intexsoft/webshop/orderservice/dto/order/OrderCreateDto.java:
```diff
@@ -0,0 +1,42 @@
+    Long userId;
+    @NotBlank
+    @ValueOfEnum(enumClass = PaymentMethod.class)
+    String paymentMethod;
```
Мы можем просто принимать параметр с типом PaymentMethod

MSurmach replied:
Да можем. Тут я вкладывал смысл в валидацию самого переданного значения, чтобы соответствовало enum. Либо можно убрать валидацию мою кастомную и перехватывать exception в случае неправильной строки в теле запроса. Exception  такой :`HttpMessageNotReadableException: JSON parse error: Cannot deserialize value of type `com.intexsoft.webshop.orderservice.model.enums.PaymentMethod` from String "CARD78": not one of the values accepted for Enum class: [CARD, CASH]]`

Tyomich4736 replied:
Да, давай лучше перехватим такой Exception и преобразуем его в красивый респонс

MSurmach replied:
Удалил кастомную валидацию, принимаю сразу PaymentMethod  в dto. Добавил handler на JsonMappingException (более узкий чем HttpMessageNotReadableException и инфы об исключении поменьше неинтереной для пользователя).

Tyomich4736 commented on user-service/src/main/java/com/intexsoft/webshop/userservice/mapper/UserEventMapper.java:
```diff
@@ -0,0 +1,15 @@
+public interface UserEventMapper {
+    @Mapping(target = "orderId", source = "checkOrderUserCommand.orderId")
+    @Mapping(target = "userId", source = "checkOrderUserCommand.userId")
+    UserCheckedEvent toUserCheckedEvent(boolean checkResult, CheckOrderUserCommand checkOrderUserCommand);
```
Не очевидно куда присваивается checkResult. Обычно в мапперах для всех полей указывается @Mapping, даже когда поля нужно игнорировать. Это относится и ко всем остальным мапперам

MSurmach replied:
Поправил все мапперы, указал явный @Mapping для всех target полей

Tyomich4736 commented on order-orchestrator/src/main/java/com/intexsoft/webshop/orderorchestrator/producer/AbstractOrderOrchestratorCommandProducer.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+
+@Slf4j
+public abstract class AbstractOrderOrchestratorCommandProducer implements OrderOrchestratorCommandProducer {
```
Вряд ли у нас подразумевается наличие какого-то особенного функционала для какого-то продюсера. Проще просто сделать один сервис с четырьмя методами для отправки сообщения

MSurmach replied:
Заменил на одного продюсера `OrderOrchestratorCommandProducerImpl` в котором четыре метода для каждой команды. Exchanges and routings в этом же сервисе для рассылки в каждый сервис

Tyomich4736 commented on order-orchestrator/src/main/java/com/intexsoft/webshop/orderorchestrator/producer/AbstractOrderOrchestratorCommandProducer.java:
```diff
@@ -0,0 +1,26 @@
+    protected String routingPrefix;
+
+    @Override
+    public void convertAndSendCommand(BaseCommand baseCommand, CommandType commandType) {
```
Что будет, если я вызову этот метод с baseCommand, не соответствующей commandType? Лучше сделать у baseCommand абстрактный метод (getCommandType() например), который будет переопределятся у каждой реализации и возвращать соответствующий этой реализации тип

MSurmach replied:
Сообщение отошлется, просто роутинг будет не совсем корректный со стороны продюсера, но для лисенара в сервисах роли не сыграет, так как там прописан роутинг для комманд через  # на конце и сообщение перехватывается @RabbitHandler. Тут я это сделал для передачи в routing в качестве ключа на конце.

Убрал это, в продюсере теперь методы для каждого ивента (не через родительский класс). Routing-key который конкатенируется к основному routing прописал в application.yaml как делал раньше.

Tyomich4736 commented on order-orchestrator/src/main/java/com/intexsoft/webshop/orderorchestrator/service/impl/OrderOrchestratorUserTaskProcessorImpl.java:
```diff
@@ -0,0 +1,85 @@
+                .processInstanceBusinessKey(processBusinessKey)
+                .taskDefinitionKey(taskDefKey);
+        Task task;
+        while (Objects.isNull(task = taskQuery.singleResult())) {
```
Есть риск войти в бесконечный цикл, нужно придумать другое решение и предусмотреть такую ситуацию

MSurmach replied:
сделал через свой ретрай в`OrderOrchestratorUserTaskProcessorImpl.class`. 12 попыток, 3 сек задержка между ними.

MSurmach committed:
Changes due to code review

MSurmach committed:
Correct product mapper in product-service

MSurmach committed:
Minor changes in all mappers
- remove ignoring unmappedTargetPolicy from all mappers
- show explicitly all autogenerated intermediate mapping methods

MSurmach committed:
Refactor entity models in product-service
- remove @Data lombok annotation (unnecessary initialization of child lazy entities)
- replace to @Getter @Setter @ToString instead of @Data
- add custom equals() and hashcode() methods based on hibernate proxies.

MSurmach committed:
Add product specification to find all products in one sql query

MSurmach committed:
Delete custom validation of PaymentMethod, instead add handler for JsonMappingException, that translate it into service exception response
